packages/devtools/lib/src/timeline/cpu_profile_protocol.dart

```diff
@@ -98,3 +99,107 @@ class CpuProfileProtocol {
     return name;
   }
 }
+
+/// Protocol for processing a [CpuStackFrame] and converting it to a bottom-up
+/// representation of the CPU profile.
+class BottomUpProfileProtocol {
+  List<CpuStackFrame> processData(CpuStackFrame stackFrame) {
+    final List<CpuStackFrame> bottomUpRoots = getRoots(stackFrame, null, []);
+
+    // Set the bottom up sample counts for each sample.
+    bottomUpRoots.forEach(cascadeSampleCounts);
+
+    // Merge samples when possible starting at the root (the leaf node of the
+    // original CPU sample).
+    mergeRoots(bottomUpRoots);
+
+    return bottomUpRoots;
+  }
+
+  /// Returns the roots for a bottom up representation of a CpuStackFrame node.
+  ///
+  /// Each root is a leaf from the original CpuStackFrame tree, and its children
+  /// will be the reverse call stack of the original sample. The stack frames
+  /// returned will not be merged to combine common roots, and the sample counts
+  /// will not reflect the bottom up sample counts. These steps will occur later
+  /// in the bottom-up conversion process.
+  @visibleForTesting
+  List<CpuStackFrame> getRoots(
+    CpuStackFrame node,
+    CpuStackFrame currentBottomUpRoot,
+    List<CpuStackFrame> bottomUpRoots,
+  ) {
+    final copy = node.shallowCopy()
+      ..exclusiveSampleCount = node.exclusiveSampleCount;
+
+    if (currentBottomUpRoot != null) {
+      copy.addChild(currentBottomUpRoot.deepCopy());
+    }
+
+    // [copy] is the new root of the bottom up call stack.
+    currentBottomUpRoot = copy;
+
+    if (node.exclusiveSampleCount > 0) {
+      // This node is a leaf node, meaning it is a bottom up root.
+      bottomUpRoots.add(currentBottomUpRoot);
+    }
+    for (CpuStackFrame child in node.children) {
+      getRoots(child, currentBottomUpRoot, bottomUpRoots);
+    }
+    return bottomUpRoots;
+  }
+
+  /// Sets sample counts of [stackFrame] and all children to
+  /// [exclusiveSampleCount].
+  ///
+  /// This is necessary for the transformation of a [CpuStackFrame] to its
+  /// bottom-up representation. This is an intermediate step between
+  /// [getRoots] and [mergeRoots].
+  @visibleForTesting
+  void cascadeSampleCounts(CpuStackFrame stackFrame) {
+    stackFrame.inclusiveSampleCount = stackFrame.exclusiveSampleCount;
+    for (CpuStackFrame child in stackFrame.children) {
+      child.exclusiveSampleCount = stackFrame.exclusiveSampleCount;
+      cascadeSampleCounts(child);
+    }
+  }
+
+  /// Merges bottom up roots that share a common call stack (starting at the
+  /// root).
+  ///
+  /// Ex. C               C                     C
+  ///      -> B             -> B        -->      -> B
```

Comment:
Crazy nit:
You can include unicode line characters in dart code
so
Write
https://www.fileformat.info/info/unicode/block/box_drawing/list.htm
C 
┗ B
    ┗ A

C 
┗ B
    ┣ A
    ┗ A